src/compressed_tensors/transforms/base.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, Union
+
+import torch
+from compressed_tensors.registry.registry import RegistryMixin
+from compressed_tensors.transforms.utils import apply_matrix_transform
+
+
+__all__ = ["Transforms"]
+
+
+# TODO: We don't need to save all the __call__ args for serialization or even have
+# them defined by a recipe. Some of them, such as if the transformation should be the
+# first or second matirx in torch.matmul depending on dimensions, can be inferred
+# by the layer time likely.
+
+MATIRX_TRANSFORMS = ["matrix-mul", "hadamard", "random-hadamard"]
+
+
+class Transforms(RegistryMixin):
+    def __new__(
+        cls,
+        transform: torch.Tensor,
+        device: Optional[Union[str, torch.device]] = "cuda",
+        dtype: Optional[torch.dtype] = torch.bfloat16,
+        *args,
+        **kwargs,
+    ):
+        """
+        Base class for setting up transforms. The registry creates transforms
+        as parameters which can be attached to modules.
+
+        import torch
+
+        size = 1024
+        dtype = torch.bfloat16
+        module = torch.nn.Linear(size, size)
+
+        hadamard_transform = Transforms.load_from_registry(
+            "random_hadamard", size=size, dtype=dtype
+        )
+        hadamard_apply = Transforms.fetch_apply("random_hadamard")
+        module.weight_transform = hadamard_transform
+
+        transformed_output = hadamard_apply(input_tensor=module.weight,
+            transform=moduel.weight_transform)
+
+        hadamard_inverse = Transforms.fetch_inverse_apply("random_hadamard")
+        original_weight = hadamard_inverse(input_tensor=transformed_output,
+            transform=model.weight_trainsform,
+            transpose=True)
```

Comment:
re-posting a comment from the other PR: I don't understand the motivation for `load_from_registry`, `fetch_apply` and`fetch_inverse_apply`. Placing some pseudocode, i think a simple object-oriented approach can clean this up to
```suggestion
        hadamard_transform = Transforms.load_from_registry(
            "random_hadamard", size=size, dtype=dtype
        )
        module.weight_transform = hadamard_transform

        transformed_output = hadamard_transform.apply(input_tensor=module.weight,
            transform=moduel.weight_transform)

        original_weight = hadamard_transform.inverse_apply(input_tensor=transformed_output,
            transform=model.weight_trainsform,
            transpose=True)
```

Reply:
1. `load_from_registry` is the standard path to fetch objects registered using our RegistryMixin
2. apply/inverse_apply are static functions which we just need access to during runtime. The fetching methods are ways to easily fetch them but happy to talk about alternatives. We don't need to be carrying around more than the transform and call_args, especially since there's strong commonality among the apply methods for each registered transform.

Reply:
ok sounds good, my thinking is just that whenever we have data (the matrix) and a method (apply/inverse_apply) we can wrap it in a class to avoid multiple references, but maybe we need direct access to the `torch.nn.Parameter`?

Reply:
Not a blocker, but in terms of usage I feel the object oriented approach from Brian is easier, since it does not require the users to fetch a function by name again and again. Thoughts?

Reply:
Resolved as per offline discussion